kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/GCSToBQLoadRunnable.java

```diff
@@ -0,0 +1,232 @@
+        // initialize maps, if we haven't seen this table before.
+        tableToURIs.put(table, new ArrayList<>());
+        tableToCurrentLoadSize.put(table, 0L);
+      }
+
+      long newSize = tableToCurrentLoadSize.get(table) + blob.getSize();
+      // if this file does not cause us to exceed our per-request quota limits...
+      if (newSize <= MAX_LOAD_SIZE_B && tableToURIs.get(table).size() < FILE_LOAD_LIMIT) {
+        // ...add the file (and update the load size)
+        tableToURIs.get(table).add(blob);
+        tableToCurrentLoadSize.put(table, newSize);
+      }
+    }
+    return tableToURIs;
+  }
+
+  /**
+   * Given a blob, return the {@link TableId} this blob should be inserted into.
+   * @param blob the blob
+   * @return the TableId this data should be loaded into, or null if we could not tell what
+   *         table it should be loaded into.
+   */
+  private TableId getTableFromBlob(Blob blob) {
+    if (blob.getMetadata() == null
+        || blob.getMetadata().get(GCSToBQWriter.GCS_METADATA_TABLE_KEY) == null) {
+      logger.error("Found blob {}\\{} with no metadata.", blob.getBucket(), blob.getName());
+      return null;
+    }
+
+    String serializedTableId = blob.getMetadata().get(GCSToBQWriter.GCS_METADATA_TABLE_KEY);
+    Matcher matcher = METADATA_TABLE_PATTERN.matcher(serializedTableId);
+    String project = matcher.group("project");
+    String dataset = matcher.group("dataset");
+    String table = matcher.group("table");
+
+    if (project == null || dataset == null || table == null) {
+      logger.error("Found blob `{}/{}` without parsable table metadata.",
+                   blob.getBucket(), blob.getName());
+      return null;
+    }
+
+    return TableId.of(project, dataset, table);
+  }
+
+  private Map<Job, List<Blob>> triggerBigQueryLoadJobs(Map<TableId, List<Blob>> tablesToBlobs) {
+    Map<Job, List<Blob>> newJobs = new HashMap<>(tablesToBlobs.size());
+    for (Map.Entry<TableId, List<Blob>> entry : tablesToBlobs.entrySet()) {
+      newJobs.put(triggerBigQueryLoadJob(entry.getKey(), entry.getValue()), entry.getValue());
+    }
+    return newJobs;
+  }
+
+  private Job triggerBigQueryLoadJob(TableId table, List<Blob> blobs) {
+    List<String> uris = blobsToURIs(blobs);
+    // create job load configuration
+    LoadJobConfiguration loadJobConfiguration =
+        LoadJobConfiguration.newBuilder(table, uris)
+            .setFormatOptions(FormatOptions.json())
+            .setCreateDisposition(JobInfo.CreateDisposition.CREATE_IF_NEEDED)
+            .build();
+    // create and return the job.
+    Job job = bigQuery.create(JobInfo.of(loadJobConfiguration));
+    // update active jobs and claimed blobs.
+    activeJobs.put(job, blobs);
+    claimedBlobs.addAll(blobs);
+    logger.info("Triggered load job for table {} with {} blobs.", table, blobs.size());
+    return job;
+  }
+
+  private List<String> blobsToURIs(List<Blob> blobs) {
+    List<String> uris = new ArrayList<>(blobs.size());
```

Comment:
The functional programmer in me really just wants to use blobs.stream().map().collect() for this. Since it's Java, not sure if it's going to look better or not, but worth checking.

Reply:
DOOO ITTTTT

Reply:
I also wanted to, but I'm just really non-fluent in functional java. So, sure :)